# Bio5488 SP2025: Assignment 11

### Massively Parallel 5' Splice Site Analysis

**Due Date:** Friday, 4/11, 11:59 PM  

**Last Modified:** 3/31/2025

---
## Overview

The objective of this assignment is to gain proficiency in analyzing functional genetic assays. In this case, we will use a portion of a published dataset 
(Wong, et al., “Quantitative Activity Profile and Context Dependence of All Human 5' Splice Sites”, Mol. Cell. 2018 Sep 20;71(6):1012-1026.e3 at https://pubmed.ncbi.nlm.nih.gov/30174293/).
But we will imagine that the data is part of your hypothetical thesis project studying 5’ splice site mutations in BRCA2. The processed data was obtained via the MAVE-NN package, 
as described here: https://mavenn.readthedocs.io/en/latest/datasets/dataset_mpsa.html. If you want to see the Python code used to process the raw data, visit that link.

Mutations that affect splicing are among the most common non-coding pathogenic mutations. But it is difficult to predict whether a mutation in a splice site will alter splicing, because human genes contain many noncanonical splice sites that lack the consensus sequence yet are spliced effectively. Intron 17 of BRCA2 contains an unusual ‘GC’ 5’ splice site sequence CAG/GCAAGUUU, which has a C at the +2 position rather than the more typical U. Pathogenic mutations at positions near the GC have been found in cancer samples, but it is unclear what the rules are that determine whether a splice site variant is functional.  

Imagine that you are studying splicing mutations in BRCA2. To discover what sequence patterns are necessary for functional splicing, you conducted a massively parallel splicing assay (MPSA) by testing all 32,768 possible sequences with either GU or GC at the +1 and +2 positions. (See Figs 1 and 2 for illustrations of splice site positions.)  9-mer NNN/GYNNNN sequences were placed at the beginning of intron 17 of a BRCA2 minigene. 20 nt barcodes were used to identify the upstream splice sequence. The MPSA measured whether exon 17 was included in the reporter transcript. After conducting the experiment, you calculated percent spliced in (PSI) as your quantitative measurement of function. PSI for a particular sequence is the percentage of detected barcode reads that included exon 17, and it is normalized to the consensus splice site. (If a splice site variant was correctly spliced as frequently as the consensus site, then its PSI = 100%.) For details, see the paper.  

--- 
## Reminders
Always remember to use an interactive job to run command-line tools (e.g., python, bowtie2, bcftools)
on the RIS. Once you submit an interactive job, you must activate your conda environment to use the
tools and dependencies within the docker image. 

```bash
$ bsub -Is -q workshop-interactive -G compute-workshop -a 'docker(takinwe1/bio5488:0.0)'
/bin/bash
$ conda activate bio5488
```
---
## Introduction
Copy all files from the hw directory to your home driectory or your work directory. Make sure you move your final to the submissions folder! 
You are provided with the following files: 
- `mpsa_analysis.py` Python script to complete
- `mpsa_data.csv` MPSA data file
- `brca2_mutations.txt` Text file of sequences of pathogenic BRCA2 splite site mutations

Use these files, the material in this document, the TA slides, and the commented instructions in mpsa_analysis.py, to complete the code and answer the questions.  
Because this is an exploratory analysis, I recommend trying out your code in an interactive environment such as jupyter notebook. Then, paste your completed code into mpsa_analysis.py and run it to complete the assignment.  

**NOTE**: This assignment requires the `numpy` and `pandas` python packages

---

### Task 1: Complete helper function
Complete ss_to_string as instructed in the script. This function converts 5’ splice site coordinates to python string coordinates. Your sequences are represented in python as strings but you consider splice site coordinates in your analysis (See slides for visual). The function ss_to_string will ensure that you can easily convert coordinates without careless math mistakes.  

### Task 2: Examine the distribution of PSI values. 
