# Bio5488 SP2025: Assignment 11

### Massively Parallel 5' Splice Site Analysis

**Due Date:** Friday, 4/11, 11:59 PM  

**Last Modified:** 3/31/2025

---
## Overview

The objective of this assignment is to gain proficiency in analyzing functional genetic assays. In this case, we will use a portion of a published dataset 
(Wong, et al., “Quantitative Activity Profile and Context Dependence of All Human 5' Splice Sites”, Mol. Cell. 2018 Sep 20;71(6):1012-1026.e3 at https://pubmed.ncbi.nlm.nih.gov/30174293/).
But we will imagine that the data is part of your hypothetical thesis project studying 5’ splice site mutations in BRCA2. The processed data was obtained via the MAVE-NN package, 
as described here: https://mavenn.readthedocs.io/en/latest/datasets/dataset_mpsa.html. If you want to see the Python code used to process the raw data, visit that link.

Mutations that affect splicing are among the most common non-coding pathogenic mutations. But it is difficult to predict whether a mutation in a splice site will alter splicing, because human genes contain many noncanonical splice sites that lack the consensus sequence yet are spliced effectively. Intron 17 of BRCA2 contains an unusual ‘GC’ 5’ splice site sequence CAG/GCAAGUUU, which has a C at the +2 position rather than the more typical U. Pathogenic mutations at positions near the GC have been found in cancer samples, but it is unclear what the rules are that determine whether a splice site variant is functional.  


Refer to https://samtools.github.io/bcftools/howtos/query.html for tutorials on bcftools

Always remember to use an interactive job to run command-line tools (e.g., python, bowtie2, bcftools)
on the RIS. Once you submit an interactive job, you must activate your conda environment to use the
tools and dependencies within the docker image. 

```bash
$ bsub -Is -q workshop-interactive -G compute-workshop -a 'docker(takinwe1/bio5488:0.0)'
/bin/bash
$ conda activate bio5488
```
---
